### PR TITLE
refactor: redesign bridge code base

### PR DESCRIPTION
**Pull request by andycall**
Complete re-design the code base of bridge, fix hard-core problems and improve the performance.

Note: this PR removed 95% of original bridge's code.

The hard-core problems in bridge:

1. All implements are mixed with QuickJS API, memory management are messed, no one knows how many retaining bugs in there.
2. Al the binding glue code are written by hand, it wasted too much time.
3. The DOM tree are implemented by QuickJS data structures, it's not efficient. 

The new design:

1. [x] Split out the code base to `core` and `bindings` directory. core aims to DOM and Web API, bindings aims to all interactions to QuickJS API. 
2. [x] Manage QuickJS JSValue reference count with stack.
3. [x] Use code generator to generate all API glue code. 
6. [x] Make all initialize data to static and read-only.
7. [x] re-design DOM API
    1. [x] EventTarget
    2. [x] Node
    3. [x] ContainerNode
    5. [x] Element
    8. [x] Document
    9. [x] Attr
    10. [x] CharacterData
    11. [x] DocumentFragment
    12. [x] Text
    13. [x] NodeList
    14. [x] CSSStyleDeclaration
    15. [x] CustomElement
    16. [x] CustomEvent
    17. [x] DOM Selector
18. [x] Make more useful string statically and pre-allocated at build stage. 

Design Notes:

## Type kind conversion with TypeScript and C++.

Based on the type define in TypeScript typings and generated the corresponding C++ call code. The call code are formed by C++ `Converter` class, responsible to easy convert any JS kind to C++ kind.

| JS                                             | C++                                        |
| ---------------------------------------------- | ------------------------------------------ |
| double                                         | TSDouble                                   |
| int64                                          | TSInt64                                    |
| boolean                                        | TSBool                                     |
| double?                                        | `TSOptional<TSDouble>`                |
| double \| null                                 | `TSNullable<TSDouble>`                       |
| double[]                                       | `TSSequence<TSDouble>`                       |
| type BlobPart = { name: string, value: double} | TSBlobPart // 自定义类型需要独立实现转换器 |
| any                                            | TSAny                                      |

Example:

```c++

template <>
struct Converter<TSDouble> : public ConverterBase<TSDouble> {
  static ImplType FromValue(JSContext* ctx, JSValue value) {
    assert(!JS_IsException(value));
    double v;
    JS_ToFloat64(ctx, &v, value);
    return v;
  }

  static JSValue ToValue(JSContext* ctx, double v) { return JS_NewFloat64(ctx, v); }
};
```

image

##  How properties and methods defined

JS code use prototype to access native defined properties and methods.

### CODE REVIEW

**Comment by andycall** [comment]
image

Append 10000 div elements
```javascript
let container = document.createElement('div');
for(let i = 0; i < 1000; i ++) {
    let child = document.createElement('div');
    for(let j = 0; j < 10; j ++) {
        let span = document.createElement('span');
        let text = document.createElement('helloworld');
        span.appendChild(text);
        child.appendChild(span);
    }
    container.appendChild(child);
}
```

**Before**: 50.7ms
**After**: 29.6 ms

**Comment by andycall** [comment]
image

Insert 1000 DIV on the top.

```javascript
  let container = document.createElement('div');
  let child = document.createElement('div');
  let span = document.createElement('span');
  let text = document.createElement('helloworld');
  span.appendChild(text);
  container.appendChild(child);

  for (let i = 0; i < 1000; i ++) {
    let span = document.createElement('span');
    let text = document.createElement('helloworld');
    span.appendChild(text);
    child.insertBefore(span, child.firstChild);
  }
```

**Before**:  10.8ms
**After**: 2.62ms

File: bridge/polyfill/src/index.ts [hunk 1]
```diff
@@ -34,13 +30,12 @@ defineGlobalProperty('asyncStorage', asyncStorage);
...
 defineGlobalProperty('webf', webf);
-defineGlobalProperty('ErrorEvent', ErrorEvent);
 
 function defineGlobalProperty(key: string, value: any, isEnumerable: boolean = true) {
   Object.defineProperty(globalThis, key, {
-    value: value,
-    enumerable: isEnumerable,
-    writable: true,
-    configurable: true
+      value: value,
```
Comment (XGHeaven) [inline bridge/polyfill/src/index.ts:36]: indent

File: bridge/polyfill/src/location.ts [hunk 2]
```diff
@@ -3,18 +3,16 @@
 * Copyright (C) 2022-present The WebF authors. All rights reserved.
 */
 
-import { URL } from './url';
-import { webf } from './webf';
+import {URL} from './url';
```
Comment (XGHeaven) [inline bridge/polyfill/src/location.ts:6]: code style: space around `URL`.

File: bridge/polyfill/src/location.ts [hunk 3]
```diff
@@ -53,7 +51,7 @@ export const location = {
     };
   },
   get reload() {
-    return bindReload;
+      return webfLocationReload.bind(this);
```
Comment (XGHeaven) [inline bridge/polyfill/src/location.ts:54]: indent

File: integration_tests/scripts/core_integration_starter.js [hunk 4]
```diff
@@ -28,23 +28,28 @@ function startIntegrationTest() {
...
       'skia-deterministic-rendering': true,
-      KRAKEN_TEST_DIR: path.join(__dirname, '../')
+      WEBF_TEST_DIR: path.join(__dirname, '../')
     },
     cwd: process.cwd(),
     stdio: 'inherit'
   });
 
   tester.on('close', (code) => {
+    console.log(code);
```
Comment (XGHeaven) [inline integration_tests/scripts/core_integration_starter.js:41]: Remove debug code

**Review by XGHeaven — CHANGES_REQUESTED** [review]
- `bridge/third_party/quickjs/quickjs.c` file is empty, remove it.
- The integration snapshot files about `hsl`/`hsla` should not be added. Maybe rebase the main branch and rerun to fix it.

(PS: 实在是看不懂，只能 re 出一些无关紧要的东西了)

File: bridge/CMakeLists.txt [hunk 5]
```diff
@@ -7,9 +7,9 @@ set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
-if (${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
-  set(CMAKE_OSX_ARCHITECTURES "x86_64;arm64")
-endif()
+#if (${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
```
[the remark this answers is not in the document]
Comment (andycall) [reply bridge/CMakeLists.txt:10]: My mistake

**Review by devjiangzhou — APPROVED** [review]

**Pull request merged**